### PR TITLE
core: signeddata implementation for builder registration

### PR DESCRIPTION
**Pull request by ciaranmcveigh5**
- addition of signeddata implementation
- added test case
- added SignedValidatorRegistration to testutils random
- update of go-eth2-version

category: feature
ticket: #849

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -60,8 +60,8 @@ github.com/alecthomas/units v0.0.0-20190924025748-f65c72e2690d/go.mod h1:rBZYJk5
 github.com/anmitsu/go-shlex v0.0.0-20161002113705-648efa622239/go.mod h1:2FmKhYUyUczH0OGQWaF5ceTx0UBShxjsH6f8oGKYe2c=
 github.com/antihax/optional v1.0.0/go.mod h1:uupD/76wgC+ih3iEmQUL+0Ugr19nfwCT1kdvxnR2qWY=
 github.com/armon/consul-api v0.0.0-20180202201655-eb2c6b5be1b6/go.mod h1:grANhF5doyWs3UAsr3K4I6qtAmlQcZDesFNEHPZAzj8=
-github.com/attestantio/go-eth2-client v0.11.7 h1:mRCtVZYg/Tmp47mNkWKnro/Wj63+jj1vz9BWNQN8MYo=
-github.com/attestantio/go-eth2-client v0.11.7/go.mod h1:zXL/BxC0cBBhxj+tP7QG7t9Ufoa8GwQLdlbvZRd9+dM=
+github.com/attestantio/go-eth2-client v0.11.8-0.20220725205133-c05328b73d87 h1:rG3xHSJSanANf/oQSc6z7y0UMWdfVP/GZMSyCUAljQY=
```
Comment (ciaranmcveigh5) [inline go.sum:63]: Jim hasn't cut a new release yet with the validator registration changes so have updated to the latest commit

can switch to v0.11.8 when jim cuts the release

File: core/signeddata.go [hunk 2]
```diff
@@ -34,6 +34,7 @@ var (
 	_ SignedData = Signature{}
 	_ SignedData = SignedVoluntaryExit{}
 	_ SignedData = VersionedSignedBlindedBeaconBlock{}
+	_ SignedData = SignedValidatorRegistration{} // name in the builder spec SignedValidatorRegistrationV1 or SignedBuilderRegistration
```
Comment (ciaranmcveigh5) [inline core/signeddata.go:37]: Not sure what naming convention you would like for this section of the codebase ie builderRegistration or validatorRegistration 

within go-eth2-client its using the spec name validatorRegistration

File: core/signeddata.go [hunk 3]
```diff
@@ -34,6 +34,7 @@ var (
 	_ SignedData = Signature{}
 	_ SignedData = SignedVoluntaryExit{}
 	_ SignedData = VersionedSignedBlindedBeaconBlock{}
+	_ SignedData = VersionedSignedValidatorRegistration{} // name in the builder spec SignedValidatorRegistrationV1 or SignedBuilderRegistration
```
Comment (ciaranmcveigh5) [inline core/signeddata.go:37]: not sure on preferred naming for this section 

validatorRegistration or BuilderRegistration 

within go-eth2-client it follows the builder spec name validatorRegistration

Comment (corverroos) [reply core/signeddata.go:37]: aligning with `eth2api.VersionedSignedValidatorRegistration` is good I think

File: core/signeddata.go [hunk 4]
```diff
@@ -508,6 +509,107 @@ func (e *SignedVoluntaryExit) UnmarshalJSON(b []byte) error {
...
+}
+
+// clone returns a copy of the VersionedSignedValidatorRegistration.
+// It is similar to Clone that returns the SignedData interface.
+//nolint:revive // similar method names.
+func (r VersionedSignedValidatorRegistration) clone() (VersionedSignedValidatorRegistration, error) {
+	var resp VersionedSignedValidatorRegistration
+	err := cloneJSONMarshaler(r, &resp)
+	if err != nil {
+		return VersionedSignedValidatorRegistration{}, errors.Wrap(err, "clone block")
```
Comment (corverroos) [inline core/signeddata.go:534]: "clone registration"

File: core/signeddata.go [hunk 5]
```diff
@@ -508,6 +509,107 @@ func (e *SignedVoluntaryExit) UnmarshalJSON(b []byte) error {
...
+	switch r.Version {
+	case spec.BuilderVersionV1:
+		marshaller = r.VersionedSignedValidatorRegistration.V1
+	default:
+		return nil, errors.New("unknown version")
+	}
+
+	registration, err := marshaller.MarshalJSON()
+	if err != nil {
+		return nil, errors.Wrap(err, "marshal block")
```
Comment (corverroos) [inline core/signeddata.go:576]: "marshal registration"

File: core/signeddata.go [hunk 6]
```diff
@@ -508,6 +509,107 @@ func (e *SignedVoluntaryExit) UnmarshalJSON(b []byte) error {
...
+	if err := json.Unmarshal(input, &raw); err != nil {
+		return errors.Wrap(err, "unmarshal validator (builder) registration")
+	}
+
+	resp := eth2api.VersionedSignedValidatorRegistration{Version: spec.BuilderVersion(raw.Version)}
+	switch resp.Version {
+	case spec.BuilderVersionV1:
+		registration := new(eth2v1.SignedValidatorRegistration)
+		if err := json.Unmarshal(raw.Registration, &registration); err != nil {
+			return errors.Wrap(err, "unmarshal phase0")
```
Comment (corverroos) [inline core/signeddata.go:601]: "unmarshal registration"

File: go.sum (hunk 1 above)
Comment (corverroos) [reply go.sum:63]: cool

**Review by corverroos — APPROVED** [review]

**Pull request merged**